apps/api-extractor/src/generators/PackageTypingsGenerator.ts

```diff
@@ -21,34 +28,56 @@ class Entry {
    * The original name of the symbol, as exported from the module (i.e. source file)
    * containing the original TypeScript definition.
    */
-  public localName: string;
+  public readonly localName: string;
 
   /**
    * The localName, possibly renamed to ensure that all the top-level exports have unique names.
    */
-  public uniqueName: string | undefined = undefined;
+  public get uniqueName(): string | undefined {
+    return this._uniqueName;
+  }
+
+  public set uniqueName(value: string | undefined) {
+    this._uniqueName = value;
+    this._sortKey = undefined; // invalidate the cached value
+  }
 
   /**
    * The compiler symbol where this type was defined, after following any aliases.
    */
-  public followedSymbol: ts.Symbol;
+  public readonly followedSymbol: ts.Symbol;
+
+  /** {@inheritdoc IFollowAliasesResult.importPackagePath} */
+  public readonly importPackagePath: string | undefined;
+
+  /** {@inheritdoc IFollowAliasesResult.importPackageExportName} */
+  public readonly importPackageExportName: string | undefined;
 
   /**
    * If true, this entry should be emitted using the "export" keyword instead of the "declare" keyword.
    */
   public exported: boolean = false;
 
+  private _uniqueName: string | undefined = undefined;
```

Comment:
>_uniqueName [](start = 10, length = 11)

what is this for?

Reply:
@iclanton asked the same question, and I couldn't remember why.

Here's the scenario:

Suppose your exported function F returns a class C, which is NOT exported by the package.  Normally the compiler prohibits this inside a TypeScript file, but it can happen if you import the type from another source file (because the compiler operates on individual files without knowing the package entry point).

If another function G relies on another class that is also called "C", then we can end up with two different C's that are part of the API surface but not exported.  This code was supposed to handle that by renaming one of them to C_2.

This was part of my original prototype, but it's completely untested.  I'll either add a test case to this PR, or else remove the feature and reintroduce it in a later PR.

---
In reply to: [160311870](https://github.com/Microsoft/web-build-tools/pull/473#discussion_r160311870) [](ancestors = 160311870)

Reply:
See IForgottenExport_2 in api-extractor-test-01/dist/index-internal.d.ts for the test case.

---
In reply to: [160535165](https://github.com/Microsoft/web-build-tools/pull/473#discussion_r160535165) [](ancestors = 160535165,160311870)